In [3]:
import pandas as pd

In [4]:
campaign_data = {
    "Network": ["Meta Ads", "Google AC", "Applovin", "TikTok"],
    "Spend ($)": [2000, 1500, 1200, 800],
    "Installs": [1000, 800, 700, 600],
    "CPI ($)": [2.00, 1.88, 1.71, 1.33],
    "D1 Retention": [42, 38, 33, 40],
    "D3 Retention": [25, 22, 19, 20],
    "D7 ROAS (%)": [35, 45, 30, 20]
}

In [5]:
df = pd.DataFrame(campaign_data)

df= df.set_index("Network") # I set the 'Network' column as an index to make the data easier to read.

df

Spend ($)  Installs  CPI ($)  D1 Retention  D3 Retention  \
Network                                                               
Meta Ads        2000      1000     2.00            42            25   
Google AC       1500       800     1.88            38            22   
Applovin        1200       700     1.71            33            19   
TikTok           800       600     1.33            40            20   

           D7 ROAS (%)  
Network                 
Meta Ads            35  
Google AC           45  
Applovin            30  
TikTok              20

##  PART 1 — Data Analysis

### 1. Calculate the Cost per Retained User (Day 3) for each network.

In [6]:
# D3 CPRU = CPI / D3 Retention

In [7]:
# I converted the D3 Retention values in campaign_data to decimal numbers.
df['D3_Retention_Decimal'] = df['D3 Retention'] / 100

In [8]:
df['D3_CPRU ($)'] = df['CPI ($)'] / df['D3_Retention_Decimal']

df['D3_CPRU ($)']

Network
Meta Ads     8.000000
Google AC    8.545455
Applovin     9.000000
TikTok       6.650000
Name: D3_CPRU ($), dtype: float64

### 2. Rank the networks based on cost efficiency and ROI potential.

In [9]:
df

Spend ($)  Installs  CPI ($)  D1 Retention  D3 Retention  \
Network                                                               
Meta Ads        2000      1000     2.00            42            25   
Google AC       1500       800     1.88            38            22   
Applovin        1200       700     1.71            33            19   
TikTok           800       600     1.33            40            20   

           D7 ROAS (%)  D3_Retention_Decimal  D3_CPRU ($)  
Network                                                    
Meta Ads            35                  0.25     8.000000  
Google AC           45                  0.22     8.545455  
Applovin            30                  0.19     9.000000  
TikTok              20                  0.20     6.650000

In [10]:
# Cost Efficiency = > Lower is better
# ROI (Return on Investment) = Net Profit / Cost of Investment = > Higher is better

In [11]:
df_cost_efficiency = df.sort_values(by='D3_CPRU ($)', ascending=True)

print("Cost Efficiency Ranking")
print(df_cost_efficiency[['D3_CPRU ($)', 'CPI ($)']])

Cost Efficiency Ranking
           D3_CPRU ($)  CPI ($)
Network                        
TikTok        6.650000     1.33
Meta Ads      8.000000     2.00
Google AC     8.545455     1.88
Applovin      9.000000     1.71


In [12]:
df_roi_potential = df.sort_values(by='D7 ROAS (%)', ascending=False)

print("ROI Potential Ranking")
print(df_roi_potential[['D7 ROAS (%)', 'CPI ($)']])

ROI Potential Ranking
           D7 ROAS (%)  CPI ($)
Network                        
Google AC           45     1.88
Meta Ads            35     2.00
Applovin            30     1.71
TikTok              20     1.33


###  3. Based on the data, which network(s) would you scale next, pause, or re-test — and why?

In [13]:
# Average D3_CPRU

df_cost_efficiency[['D3_CPRU ($)']].mean()

D3_CPRU ($)    8.048864
dtype: float64

In [14]:
# Average D7 ROAS

df_roi_potential[['D7 ROAS (%)']].mean()

D7 ROAS (%)    32.5
dtype: float64

``` bash 

*** Google AC => Scale: The network with the highest D7 ROAS (45%). This indicates it brings in the most profitable user base. CPI = $1.88, CPRU = $8.55, and D7 ROAS = 45%

- The network generates the highest monetization (earnings) per user.
- While CPRU is slightly below average, profitability is the highest priority.

```

``` bash
*** Meta Ads => Re-Test: While D3 CPRU is better than Google's, D7 ROAS (35%) is lower. D7 ROAS = 35%, CPRU = $8.00 → seems balanced. CPI = $2.00 → slightly higher than other networks, but ROI potential appears positive.

- Considering the CPRU cost, ROAS is not bad. ROAS is 35% → has potential.
- The network performs well in terms of user quality and ROAS. However, the high CPI and moderate CPRU could be improved.

```

``` bash

*** TikTok → Re-Test: Best CPRU ($6.65) → cheapest retained user, but D7 ROAS = 20% → lowest revenue. Its lowest D7 ROAS (20%) suggests that the users it brings in are not making money.

- Recruited users are not spending money (mostly free users), which could lead to low LTV.
- However, the user acquisition cost is low.
```

``` bash

*** Applovin → Pause: It has the worst D3 CPRU ($9.00) and the lowest D3 retention rate (19%). While its ROAS is 30%, it appears to be very weak compared to other networks in terms of efficiency. It seems more efficient to redirect spending to other, more efficient networks.

- The combination of high CPRU and low ROAS is not profitable.

```


### 4. Identify 2–3 possible hypotheses to improve performance (e.g., creative direction, audience type, event optimization).

In [15]:
# The game is monetized with 'interstitial ads' and 'rewarded videos', with a small 'in-app purchase (IAP)'share.

In [16]:
df[["D7 ROAS (%)","D3_CPRU ($)", "CPI ($)" ]]

D7 ROAS (%)  D3_CPRU ($)  CPI ($)
Network                                     
Meta Ads            35     8.000000     2.00
Google AC           45     8.545455     1.88
Applovin            30     9.000000     1.71
TikTok              20     6.650000     1.33

Hypothesis 1 — Creative-matching hypothesis (TikTok):

TikTok has the best CPI despite its low D7 ROAS. Its CPRU is also low ($6.65), meaning users are attracted at a lower cost, but quality is lower.

Hypothesis

If TikTok creatives (interstitial ads) are better matched to the gaming experience (creative matching), user quality will increase, and the D7 ROAS will increase. If interstitial creative reflects real gameplay moments, users will be attracted with genuine motivation. This will increase retention and LTV.

Hypothesis 2 — Audience Optimization Hypothesis (Meta)

Meta has a good D7 ROAS, but its CPRU is slightly high. High-quality but expensive users. This may indicate that campaigns are still focused on existing audiences and the system is not sufficiently exploring less expensive users.

Hypothesis

If the campaign target audience is optimized and expanded to broader or more cost-effective lookalike/interest audiences, the system can achieve cheaper installs and the CPI will decrease.

## PART 2 — Creative Performance Deep Dive

### 1. Analyze the relationship between CTR, CVR, CPI, and ROAS.

In [17]:
performance_data = {
    'Creative': ['A – Funny Fails', 'B – Satisfying Merge', 'C – Challenge Level'],
    'CTR (%)': [3.8, 2.2, 4.5],
    'CVR (%)': [40, 45, 35],
    'CPI ($)': [2.10, 1.80, 1.60],
    'D1 Retention': ['43%', '41%', '35%'],
    'D7 ROAS (%)': [32, 38, 25]
}

In [18]:
df_meta = pd.DataFrame(performance_data)

df_meta= df_meta.set_index("Creative")

df_meta

CTR (%)  CVR (%)  CPI ($) D1 Retention  D7 ROAS (%)
Creative                                                                 
A – Funny Fails           3.8       40      2.1          43%           32
B – Satisfying Merge      2.2       45      1.8          41%           38
C – Challenge Level       4.5       35      1.6          35%           25

CTR (Click-through rate) => (Number of clicks / number of impressions) * 100, shows how attractive the ad is to the audience.
CVR (Conversion rate) => (conversions / total visitors) * 100 , shows how effective the post-click experience is at driving installs.

Creative C: Highest CTR (4.5%) → ad attracts attention, but low CVR (35%) → most clickers don't install.

Creative B: Lowest CTR (2.2%) → fewer people click, but most clickers install (45% CVR).

Creative A: Medium CTR (3.8%) and medium CVR (40%) → ad attracts sufficient attention, and most clickers install. This demonstrates a balanced performance, potentially in a good spot in terms of both user volume and quality.

Creative C: Lowest CPI (1.6) → clicks are cheap, and some install.

Creative A: Medium CTR and CVR, but the most expensive CPI (2.1) → cost-per-click and quality are slightly high.

Creative B: Medium CPI (1.8) → increases ROAS thanks to high CVR.

Creative B: Highest D7 ROAS (38%) → clickers are high-quality and spend.

Creative C: Low ROAS (25%) → clicks are cheap, but users don't spend in-game.

Linked to D1 Retention: High D1 retention → generally high ROAS.

A: 43% → ROAS 32, B: 41% → ROAS 38, C: 35% → ROAS 25

In [19]:
## Let's make a hypothetical calculation based on a user base of 10,000 people.


fake_data = {
    "Creative": ["A – Funny Fails", "B – Satisfying Merge", "C – Challenge Level"],
    "CTR": [3.8, 2.2, 4.5],      
    "CVR": [40, 45, 35],    
    "CPI": [2.1, 1.8, 1.6],       
    "D7_ROAS": [32, 38, 25]   
}

df = pd.DataFrame(fake_data)
impressions = 10000


In [20]:
df["Clicks"] = impressions * df["CTR"] / 100

df["Installs"] = df["Clicks"] * df["CVR"] / 100

df["Cost"] = df["Installs"] * df["CPI"]

df["Revenue"] = df["Cost"] * df["D7_ROAS"] / 100

In [21]:
df["Clicks"] = df["Clicks"].round(0).astype(int)
df["Installs"] = df["Installs"].round(0).astype(int)
df["Cost"] = df["Cost"].round(2)
df["Revenue"] = df["Revenue"].round(2)

# Results
df[["Creative", "Clicks", "Installs", "Cost", "Revenue"]]

Creative  Clicks  Installs   Cost  Revenue
0       A – Funny Fails     380       152  319.2   102.14
1  B – Satisfying Merge     220        99  178.2    67.72
2   C – Challenge Level     450       158  252.0    63.00

Comments:

Creative A: Revenue $102.1 → slightly high cost, but highest ROAS. Balanced performance.

Creative B: Revenue $67.7 → low cost and high ROAS, but low total user count.

Creative C: Revenue $63.2 → high user count, low cost, but very low ROAS → low profitability.

###  2. Which creative would you continue, pause, or iterate on?

Creative B -> Continue

This creative has the highest D7 ROAS (38%), high CVR (45%), and good D1 retention (41%). This means the users who click are high-quality and profitable. Therefore, it makes sense to continue and scale.

Creative A -> Iterate

This creative performs well: moderate ROAS (32%), moderate CPI ($2.1), and good D1 retention (43%). This means users are both clicking and downloading sufficiently. ROAS can be increased with new opening hooks, visual tone, or CTA changes.

Creative C -> Pause

Despite attracting attention with a high CTR (4.5%), CVR is low (35%), and D7 ROAS is low (25%). This indicates that most users are low-quality or low-spending users. It can be paused for now or retested with optimized targeting.


Continue → ROAS ve retention yüksek, karlı kreatif

Pause → ROAS düşük, maliyet yüksek, düşük kaliteli kullanıcı

Iterate → Orta performans, optimize edilip test edilebilir

###  3. Suggest two data-backed creative iteration ideas (e.g., new opening hook, visual tone, or CTA variation)

Creative Iteration Plan – Funny Fails

Current Status (Decision: Iterate)

- CTR: 3.8% → Medium
- CVR: 40% → Medium
- D1 Retention: 43% → Strong
- D7 ROAS: 32% → Medium

| #     | Iteration Idea                     | Description                                                                                      | Target Metric(s)                   | Expected Impact                                                               |
| ----- | ---------------------------------- | ------------------------------------------------------------------------------------------------ | ---------------------------------- | ----------------------------------------------------------------------------- |
| **1** | **New Opening Hook**               | Show the funniest or most unexpected “fail” moment in the first 2 seconds of the ad.             | **CTR**, **CVR**, **D1 Retention** | Captures user attention early and increases click-through and install rates.  |
| **2** | **Visual Tone & CTA Variation**    | Use brighter colors, character-focused scenes, and clearer CTAs like “Try it now and have fun!”. | **ROAS**, **CVR**                  | Boosts engagement and drives higher in-app purchases post-install.            |
| **3** | **Sound Cue Optimization** | Add funny or surprising sound effects during “fail” moments to draw attention.                   | **CTR**, **D1 Retention**          | Makes the ad more memorable and engaging, improving watch time and retention. |


## References

- https://tr-tr.facebook.com/business/help/139628083350822
- https://ads.tiktok.com/help/article/creative-center?lang=tr
- https://www.appsflyer.com/glossary/interstitials/
- https://metalla.digital/meta-ad-hooks-that-drive-conversions-in-2025/
- https://www.ideasoft.com.tr/call-to-action-cta-nedir/
- https://segwise.ai/blog/optimization-techniques-meta-ad-campaigns-mobile-gaming
- https://www.facebook.com/business/ads-guide/update/video/audience-network-native/app-installs